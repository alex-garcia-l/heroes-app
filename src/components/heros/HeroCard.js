import React from 'react';
import { Link } from 'react-router-dom';
import { getImages } from '../../helper/getImage';

export const HeroCard = ({
  id,
  superhero,
  publisher,
  alter_ego,
  first_appearance,
  characters
}) => {

  const imagePath = getImages(`./${id}.jpg`);

  return (
    <div className="col-12 col-sm-6 col-lg-4 animate__animated animate__fadeIn">
      <div className="card mb-3">
        <div className="row g-0">
          <div className="col-4">
            <img src={imagePath} className="img-fluid rounded-start" alt={superhero} />
          </div>
          <div className="col-8">
            <div className="card-body">
              <h5 className="card-title">{superhero}</h5>
              <p className="card-text">{alter_ego}</p>
              {
                (alter_ego !== characters)
                && <p className="card-text"><small className="text-muted">{characters}</small></p>
              }
              <p className="card-text"><small>{first_appearance}</small></p>
              <Link to={`/hero/${id}`}>
                Ver mÃ¡s...
              </Link>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
