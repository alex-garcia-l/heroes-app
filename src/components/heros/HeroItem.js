import React from 'react';
import { Link } from 'react-router-dom';

export const HeroItem = ({
  id,
  superhero,
  publisher,
  alter_ego,
  first_appearance,
  characters
}) => {

  const imagePath = `/images/heroes/${id}.jpg`;

  return (
    <div className="col-12 col-lg-6">
      <div className="card mb-3">
        <div className="row g-0">
          <div className="col-4">
            <img src={imagePath} className="img-fluid rounded-start" alt={superhero} />
          </div>
          <div className="col-8">
            <div className="card-body">
              <h5 className="card-title">{superhero}</h5>
              <p className="card-text">{alter_ego}</p>
              {
                (alter_ego !== characters)
                  && <p className="card-text"><small className="text-muted">{characters}</small></p>
              }
              <p className="card-text"><small>{first_appearance}</small></p>
              <Link to={`/hero/${id}`}>
                  Ver mÃ¡s...
              </Link>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
