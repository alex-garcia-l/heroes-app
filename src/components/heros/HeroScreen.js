import React, { useMemo } from 'react';
import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getHerosById } from '../../selectors/hero/getHerosById';

export const HeroScreen = () => {

  const { heroID } = useParams();
  const hero = useMemo(() => getHerosById(heroID), [heroID]);;
  const navigate = useNavigate();

  if (!hero) {
    return <Navigate to="/home" />
  }
  const {
    id,
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters
  } = hero;
  const imagePath = `/images/heroes/${id}.jpg`;

  const handleClick = (evt) => {
    evt.preventDefault();
    navigate(-1);
  }

  return (
    <div className="row mt-5">
      <div className="col-3 offset-2">
        <img src={imagePath} className="img-thumbnail animate__animated animate__fadeInLeft" alt={hero.superhero} />
      </div>

      <div className="col-5 animate__animated animate__fadeInRight">
        <h5>{superhero}</h5>
        <ul className="list-group mt-4">
          <li className="list-group-item">
            <span className="fw-bold">Alter Ego: </span>{alter_ego}
          </li>
          <li className="list-group-item">
            <span className="fw-bold">Publisher: </span>{publisher}
          </li>
          <li className="list-group-item">
            <span className="fw-bold">First Appearance: </span>{first_appearance}
          </li>
        </ul>

        <h5 className="mt-5">Characters</h5>
        <p>{characters}</p>

        <button
          className="btn btn-outline-info"
          onClick={handleClick}
        >
          Regresar
        </button>
      </div>
    </div>
  );
};
