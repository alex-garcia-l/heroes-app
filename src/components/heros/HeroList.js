import React from 'react';
import PropTypes from 'prop-types';

import { getHerosByPublisher } from '../../selectors/hero/getHerosByPublisher';
import { HeroItem } from './HeroItem';

export const HeroList = ({ publisher }) => {

  const heroes = getHerosByPublisher(publisher);

  return (
    <div className="row">
      {
        heroes.map(hero => (
          <HeroItem
            key={hero.id}
            { ...hero }
          />
        ))
      }
    </div>
  );
};

HeroList.propTypes = {
  publisher: PropTypes.string.isRequired
}
