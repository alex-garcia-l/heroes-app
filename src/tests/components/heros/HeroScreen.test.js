import { mount } from "enzyme";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { HeroScreen } from "../../../components/heros/HeroScreen";

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}));


describe('Prueba para el componente <HeroScreen />', () => {

  const text = 'No hero page';

  test('NO debe mostrar el componente', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero']}>
        <Routes>
          <Route path='hero' element={<HeroScreen />} />
          <Route path='home' element={<h1>{text}</h1>} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper.find('h1').text().trim()).toBe(text);
  });

  test('debe mostrar el hero correctamente', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/dc-batman']}>
        <Routes>
          <Route path='hero/:heroID' element={<HeroScreen />} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.row.mt-5').exists()).toBe(true);
  });

  test('debe llamar el evento del botÃ³n regresar', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/dc-batman']}>
        <Routes>
          <Route path='hero/:heroID' element={<HeroScreen />} />
        </Routes>
      </MemoryRouter>
    );

    wrapper.find('button').simulate('click', { preventDefaukt() { } });
    expect(mockNavigate).toHaveBeenCalledWith(-1);
  });

  test('debe de redireccionar si no encuentra un hero', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/dc-batman123']}>
        <Routes>
          <Route path='hero/:heroID' element={<HeroScreen />} />
          <Route path='home' element={<h1>{text}</h1>} />
        </Routes>
      </MemoryRouter>
    );

    expect(wrapper.find('h1').text().trim()).toBe(text);
  });

});
