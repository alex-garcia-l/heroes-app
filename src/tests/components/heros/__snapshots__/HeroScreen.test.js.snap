// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba para el componente <HeroScreen /> debe mostrar el hero correctamente 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-3 offset-2"
  >
    <img
      alt="Batman"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="/images/heroes/dc-batman.jpg"
    />
  </div>
  <div
    className="col-5 animate__animated animate__fadeInRight"
  >
    <h5>
      Batman
    </h5>
    <ul
      className="list-group mt-4"
    >
      <li
        className="list-group-item"
      >
        <span
          className="fw-bold"
        >
          Alter Ego: 
        </span>
        Bruce Wayne
      </li>
      <li
        className="list-group-item"
      >
        <span
          className="fw-bold"
        >
          Publisher: 
        </span>
        DC Comics
      </li>
      <li
        className="list-group-item"
      >
        <span
          className="fw-bold"
        >
          First Appearance: 
        </span>
        Detective Comics #27
      </li>
    </ul>
    <h5
      className="mt-5"
    >
      Characters
    </h5>
    <p>
      Bruce Wayne
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Regresar
    </button>
  </div>
</div>
`;
